mlflow/autogen/autogen_logger.py

```diff
@@ -0,0 +1,262 @@
+                except Exception as e:
+                    result = None
+                    self._record_exception(span, e)
+                    raise e
+                finally:
+                    self._client.end_trace(
+                        request_id=span.request_id, outputs=result, status=span.status
+                    )
+                    # Clear the state to start a new chat session
+                    self._chat_state.clear()
+            elif not root_only:
+                span = self._start_span_in_session(
+                    name=span_name or f.__name__,
+                    span_type=span_type,
+                    inputs=capture_function_input_args(f, args, kwargs),
+                )
+                try:
+                    result = f(*args, **kwargs)
+                except Exception as e:
+                    result = None
+                    self._record_exception(span, e)
+                    raise e
+                finally:
+                    self._client.end_span(
+                        request_id=span.request_id,
+                        span_id=span.span_id,
+                        outputs=result,
+                        status=span.status,
+                    )
+                    self._chat_state.pended_spans.append(span)
+            else:
+                result = f(*args, **kwargs)
+            return result
+
+        return _wrapper
+
+    def _record_exception(self, span: Span, e: Exception):
+        try:
+            span.set_status(SpanStatus(SpanStatusCode.ERROR, str(e)))
+            span.add_event(SpanEvent.from_exception(e))
+        except Exception as e:
+            _logger.warning(
+                "Failed to record exception in span.", exc_info=_logger.isEnabledFor(logging.DEBUG)
+            )
+
+    def _start_span_in_session(
+        self,
+        name: str,
+        span_type: str,
+        inputs: Dict[str, Any],
+        attributes: Optional[Dict[str, Any]] = None,
+        start_time_ns: Optional[int] = None,
+    ) -> Span:
+        """
+        Start a span in the current chat session.
+        """
+        if self._chat_state.session_span is None:
+            _logger.warning("Failed to start span. No active chat session.")
+            return NoOpSpan()
+
+        return self._client.start_span(
+            request_id=self._chat_state.session_span.request_id,
+            # Tentatively set the parent ID to the session root span, because we
+            # cannot create a span without a parent span (otherwise it will start
+            # a new trace). The actual parent will be determined once the chat
+            # message is received.
+            parent_id=self._chat_state.session_span.span_id,
+            name=name,
+            span_type=span_type,
+            inputs=inputs,
+            attributes=attributes,
+            start_time_ns=start_time_ns,
+        )
+
+    def log_event(self, source: Union[str, Agent], name: str, **kwargs: Dict[str, Any]):
```

Comment:
Just to confirm: we have events as a property on spans. Should `log_event()` populate an event on an existing span, or should it create a new span?

Reply:
It looks like this is intending to mirror the concept of autogen's "event logging" to record the responses that agents received when calling functions. 
https://github.com/microsoft/autogen/blob/972b4ed024fbf5657467196832017ac9380da880/autogen/agentchat/conversable_agent.py#L840-L852
The naming here could be changed to make it clear that we're not capturing exception events as part of an MLflow span schema entry

Reply:
Autogen instrumentation system does not have a concept of span so everything is passed as an "event" to the logger. I know the naming is confusing, but this is override of the basic logger interface so we cannot change it.